server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs

```diff
@@ -48,43 +52,69 @@ import           Hasura.RQL.Types
-  , gqRespErrors :: Maybe [OJ.Value]
-  } deriving (Show, Eq)
-
-parseGQRespValue :: OJ.Value -> Either String GQRespValue
-parseGQRespValue =
-  \case
-    OJ.Object obj -> do
-      gqRespData <-
-        case OJ.lookup "data" obj of
-          Nothing -> pure Nothing
-          Just (OJ.Object dobj) -> pure (Just dobj)
-          Just _ -> Left "expected object for GraphQL data response"
-      gqRespErrors <-
-        case OJ.lookup "errors" obj of
-          Nothing -> pure Nothing
-          Just (OJ.Array vec) -> pure (Just (toList vec))
-          Just _ -> Left "expected array for GraphQL error response"
-      pure (GQRespValue {gqRespData, gqRespErrors})
-    _ -> Left "expected object for GraphQL response"
-
-gqrespValueToValue :: GQRespValue -> OJ.Value
-gqrespValueToValue (GQRespValue mdata' merrors) =
-  OJ.Object
-    (OJ.fromList
-       (concat
-          [ [("data", OJ.Object data') | Just data' <- [mdata']]
-          , [ ("errors", OJ.Array (V.fromList errors))
-            | Just errors <- [merrors]
-            ]
-          ]))
-
-data GQJoinError =  GQJoinError !T.Text
-  deriving (Show, Eq)
-
+  { _gqRespData   :: OJ.Object
+  -- ^ 'OJ.empty' (corresponding to the invalid `"data": {}`) indicates an error.
+  , _gqRespErrors :: VB.Builder OJ.Value
+  -- ^ An 'OJ.Array', but with efficient cons and concatenation. Null indicates
+  -- query success.
+  }
+-- TODO consider unifying this type with Hasura.GraphQL.Transport.HTTP.Protocol.
+
+makeLenses ''GQRespValue
+
+parseGQRespValue :: EncJSON -> Either String GQRespValue
+parseGQRespValue = OJ.eitherDecode . encJToLBS >=> \case
+  OJ.Object obj -> do
+    _gqRespData <-
+      case OJ.lookup "data" obj of
+        -- "an error was encountered before execution began":
+        Nothing -> pure OJ.empty
+        -- "an error was encountered during the execution that prevented a valid response":
+        Just OJ.Null -> pure OJ.empty
+        Just (OJ.Object dobj) -> pure dobj
+        Just _ -> Left "expected object or null for GraphQL data response"
+    _gqRespErrors <-
+      case OJ.lookup "errors" obj of
+        Nothing -> pure VB.empty
+        Just (OJ.Array vec) -> pure $ VB.vector vec
+        Just _ -> Left "expected array for GraphQL error response"
+    pure (GQRespValue {_gqRespData, _gqRespErrors})
+  _ -> Left "expected object for GraphQL response"
+
+encodeGQRespValue :: GQRespValue -> EncJSON
+encodeGQRespValue GQRespValue{..} = OJ.toEncJSON $ OJ.Object $ OJ.fromList $
+  -- "If the data entry in the response is not present, the errors entry in the
+  -- response must not be empty. It must contain at least one error. "
+  if _gqRespData == OJ.empty && V.null gqRespErrorsV
+    then
+      let msg = "Somehow did not accumulate any errors or data from graphql queries"
+       in [("errors", OJ.Array $ V.singleton $ gQJoinErrorToValue msg)]
+    else
+      ("data", dataOrNull) :
+      [("errors", OJ.Array gqRespErrorsV) | not $ V.null gqRespErrorsV ]
```

Comment:
This is like an inline guard?

Reply:
It's just list comprehension syntax, like the idiom you used before but with no list iterator portion thingy

Reply:
ah right right :sweat_smile: